manuals/vsmbrinfo.1
.Dd June 26, 2026
.Dt VSMBRINFO 1
.Os
.Sh NAME
.Nm vsmbrinfo
.Nd determines information about a Master Boot Record (MBR) volume system
.Sh SYNOPSIS
.Nm vsmbrinfo
.Op Fl b Ar bytes_per_sector
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm vsmbrinfo
is a utility to determine information about a Master Boot Record (MBR) volume \
system
.Pp
.Nm vsmbrinfo
is part of the
.Nm libvsmbr
package.
.Nm libvsmbr
is a library to access the Master Boot Record (MBR) volume system format
.Pp
.Ar source
is the source image.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl b Ar bytes_per_sector
specify the number of bytes per sector (default is 512).
Use this to override the automatic bytes per sector detection
.It Fl h
shows this help
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# vsmbrinfo /dev/sda
vsmbrinfo 20260626
.sp
Master Boot Record (MBR) information:
	Disk identity		: 0x53f5a6ee (1408607982)
	Bytes per sector	: 512
	Number of partitions	: 2
.sp
Partition: 1
	Type			: 0x83 (Linux)
	Offset			: 512 (0x00000200)
	Size			: 179200
.sp
Partition: 2
	Type			: 0x83 (Linux)
	Offset			: 180224 (0x0002c000)
	Size			: 1294336
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libvsmbr/issues
.Sh COPYRIGHT
Copyright (C) 2010-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO:
* add check for partition offset > volume size

20160108
* see `git log' for more recent change log
* 2016 update

20150105
* 2015 update

20141226
* removed README.macosx

20140201
* 2014 update
* updated dependencies

20130923
* 2013 update
* updated dependencies

20121215
* worked on stand-alone version

20121214
* created stand-alone version based on libsmal



vsmbrtools/vsmbrinfo.c
/*
 * Shows information obtained from a Master Boot Record (MBR) volume system.
 *
 * Copyright (C) 2010-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "info_handle.h"
#include "vsmbrtools_getopt.h"
#include "vsmbrtools_libcerror.h"
#include "vsmbrtools_libclocale.h"
#include "vsmbrtools_libcnotify.h"
#include "vsmbrtools_libvsmbr.h"
#include "vsmbrtools_output.h"
#include "vsmbrtools_signal.h"
#include "vsmbrtools_unused.h"

info_handle_t *vsmbrinfo_info_handle = NULL;
int vsmbrinfo_abort                  = 0;

/* Signal handler for vsmbrinfo
 */
void vsmbrinfo_signal_handler(
      vsmbrtools_signal_t signal VSMBRTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "vsmbrinfo_signal_handler";

	VSMBRTOOLS_UNREFERENCED_PARAMETER( signal )

	vsmbrinfo_abort = 1;

	if( vsmbrinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     vsmbrinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use vsmbrinfo to determine information about a Master Boot Record (MBR) volume system.";

	vsmbrtools_option_t options[ ] = {
		{ 'b', "bytes_per_sector", "specify the number of bytes per sector (default is 512). Use this to override the automatic bytes per sector detection" },
		{ 'h', NULL, "shows this help" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source image" },
	};
	system_character_t options_string[ 32 ];

	libvsmbr_error_t *error                     = NULL;
	system_character_t *option_bytes_per_sector = NULL;
	system_character_t *source                  = NULL;
	char *program                               = "vsmbrinfo";
	system_integer_t option                     = 0;
	int number_of_options                       = (int) ( sizeof( options ) / sizeof( vsmbrtools_option_t ) );
	int result                                  = 0;
	int verbose                                 = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "vsmbrtools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( vsmbrtools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	vsmbrtools_output_version_fprint(
	 stdout,
	 program );

	if( vsmbrtools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = vsmbrtools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				vsmbrtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'b':
				option_bytes_per_sector = optarg;

				break;

			case (system_integer_t) 'h':
				vsmbrtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				vsmbrtools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source image.\n" );

		vsmbrtools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libvsmbr_notify_set_stream(
	 stderr,
	 NULL );
	libvsmbr_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &vsmbrinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
#if defined( __clang_analyzer__ )
	__builtin_assume( vsmbrinfo_info_handle != NULL );
#endif
	if( option_bytes_per_sector != NULL )
	{
		result = info_handle_set_bytes_per_sector(
			  vsmbrinfo_info_handle,
			  option_bytes_per_sector,
			  &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set bytes per sector.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported bytes per sector defaulting to: %" PRIu32 ".\n",
			 vsmbrinfo_info_handle->bytes_per_sector );
		}
	}
	if( info_handle_open_input(
	     vsmbrinfo_info_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open: %" PRIs_SYSTEM ".\n",
		 source );

		goto on_error;
	}
	if( info_handle_partitions_fprint(
	     vsmbrinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print partitions information.\n" );

		goto on_error;
	}
	if( info_handle_close_input(
	     vsmbrinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &vsmbrinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( vsmbrinfo_info_handle != NULL )
	{
		info_handle_free(
		 &vsmbrinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

